cpp/src/interop/from_arrow_device.cu

```diff
@@ -0,0 +1,409 @@
+                                                    rmm::cuda_stream_view stream,
+                                                    rmm::mr::device_memory_resource* mr)
+{
+  return type.id() != type_id::EMPTY
+           ? std::move(type_dispatcher(
+               type, dispatch_to_cudf_column{}, schema, input, type, skip_mask, stream, mr))
+           : std::make_tuple<column_view, owned_columns_t>(get_empty_type_column(input->length),
+                                                           {});
+}
+
+}  // namespace
+
+unique_table_view_t from_arrow_device(ArrowSchemaView* schema,
+                                      const ArrowDeviceArray* input,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  if (input->sync_event != nullptr) {
+    cudaStreamWaitEvent(stream.value(), *reinterpret_cast<cudaEvent_t*>(input->sync_event));
+  }
+
+  std::vector<column_view> columns;
+  owned_columns_t owned_mem;
+
+  auto type = arrow_to_cudf_type(schema);
+  if (type != data_type(type_id::STRUCT)) {
+    auto [colview, owned] = get_column(schema, &input->array, type, false, stream, mr);
+    columns.push_back(colview);
+    owned_mem = std::move(owned);
+  } else {
+    std::transform(
+      input->array.children,
+      input->array.children + input->array.n_children,
+      schema->schema->children,
+      std::back_inserter(columns),
+      [&owned_mem, &stream, &mr](ArrowArray const* child, ArrowSchema const* child_schema) {
+        ArrowSchemaView view;
+        NANOARROW_THROW_NOT_OK(ArrowSchemaViewInit(&view, child_schema, nullptr));
+        auto type              = arrow_to_cudf_type(&view);
+        auto [out_view, owned] = get_column(&view, child, type, false, stream, mr);
+        if (owned_mem.empty()) {
+          owned_mem = std::move(owned);
+        } else {
+          owned_mem.insert(std::end(owned_mem),
+                           std::make_move_iterator(std::begin(owned)),
+                           std::make_move_iterator(std::end(owned)));
+        }
+        return out_view;
+      });
+  }
+
+  return unique_table_view_t{new table_view{columns}, custom_view_deleter{std::move(owned_mem)}};
+}
+
+}  // namespace detail
+
+unique_table_view_t from_arrow_device(const ArrowSchema* schema,
+                                      const ArrowDeviceArray* input,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(schema != nullptr && input != nullptr,
+               "input ArrowSchema and ArrowDeviceArray must not be NULL");
+  CUDF_EXPECTS(input->device_type == ARROW_DEVICE_CUDA ||
+                 input->device_type == ARROW_DEVICE_CUDA_HOST ||
+                 input->device_type == ARROW_DEVICE_CUDA_MANAGED,
+               "ArrowDeviceArray memory must be accessible to CUDA");
+
```

Comment:
Currently if the data is on the CPU, you would use `from_arrow` instead of `from_arrow_device` to do the conversion and copying between CPU and the GPU. A future PR will be to update  the existing `to_arrow`/`from_arrow` methods to leverage nanoarrow and the C Data interface to perform CPU/GPU copies.

Reply:
CPU memory that is already page-locked / registered with the driver would be `ARROW_DEVICE_CUDA_MANAGED` or `ARROW_DEVICE_CUDA_HOST` which should work fine in these functions.

Reply:
My suggestion was allowing the user to specify the target device type and id to allow `GPU->CPU` and `CPU->GPU`.

Reply:
I intend on enabling the `GPU->CPU` and `CPU->GPU` cases, just not in this PR. That will be a follow-up to this one which would be to replace the existing `from_arrow` and `to_arrow` APIs that use the C++ libarrow with CPU based memory.